### ISSUE TITLE
No Error Handling for forbidden socket connections / Amp should be optional

### ISSUE THREAD

**Issue by c33s** [humbug/box#166]
| Question    | Answer
| ------------| ---------------
| Box version | 3.0.0-alpha.3
| PHP version | 7.1.16
| Platform with version | Win7x64
| Github Repo | - 

running `box` on a firewalled windows system where all connections are blocked by default for all processes. if i run `box build` i get screens of error messages (i try to put a symfony4 skeleton in a phar file).

```
[ERROR] _HumbugBox5adef95163555\Amp\Process\ProcessException: Failed to connect socket #0: 10013: An attempt was made
         to access a socket in a way forbidden by its access permissions. in
         phar://C:/bin/box.phar/vendor/amphp/process/lib/Internal/Windows/SocketConnector.php:258
Stack trace:
```

if i allow the php process to create a connection the errors are gone and box is building. in comparision to box 2 the amp part has the opposite of a speed increase. what box2 has build in 63sec box3 was not able to do in over 30min. had to terminate the process, the firewall was configured to allowed php to access the local dynamic port and also the temp file `ampCB1B.tmp` was allowed to connect.

amp is a nice concept but it should be really optional, it has quite many side effects.

having to allow dynamic port access which i can't control was never i concept i liked. also having files in the temp dir, which are not firewalled is really uncool (need dynamic access). additional i have to bring my firewall in dialog mode every time i want to run box because the temp file is generated with a random name, so i have to create new rules for each file at each run.

please add a old school non-amp processing. 

<details>
 <summary>box.json.dist</summary>
 
 ```yaml
files:
    - composer.json
directories:
    - bin
    - config
    - src
    - lib
finder:
    -
      name: '*.*'
      in: vendor
      exclude:
        - .gitignore
        - .md
        - phpunit
        - Tester
        - Tests
        - tests
compactors:
    - Herrera\Box\Compactor\Json
    - Herrera\Box\Compactor\Php
compression: GZ
main: bin/puppet-enc
output: build/puppet-enc.phar
stub: true
git-commit: git_commit
git-version: git_version
git-tag: git_tag
chmod: '0755'

 ```
</details>

<details>
 <summary>Output</summary>
 
 ```bash
box build
converting to json


 [WARNING] The command "build" is deprecated. Use "compile" instead.



    ____
   / __ )____  _  __
  / __  / __ \| |/_/
 / /_/ / /_/ />  <
/_____/\____/_/|_|


Box version 3.0.0-alpha.3 build 6b7dc4a883c199c93403adbc7dfaefdd52fc9336


 // Loading the configuration file
 // "box.json".

Building the PHAR "./build/puppet-enc.phar"
? Setting replacement values
  + @git_commit@: 5f620495beb07c9304e4bd88dafa11e45224b1c5
  + @git_tag@: 1.0.1
  + @git_version@: 1.0.1
? Registering compactors
  + KevinGH\Box\Compactor\Json
  + KevinGH\Box\Compactor\Php
? Adding main file: ./bin/puppet-enc
? Adding requirements checker
? Adding binary files
    > No file found
? Adding files


 [ERROR] _HumbugBox5adef95163555\Amp\Process\ProcessException: Failed to connect socket #0: 10013: An attempt was made
         to access a socket in a way forbidden by its access permissions. in
         phar:///bin/box.phar/vendor/amphp/process/lib/Internal/Windows/SocketConnector.php:258
Stack trace:
#0
         phar:///bin/box.phar/vendor/amphp/amp/lib/Loop/NativeDriver.php(91):
         _HumbugBox5adef95163555\Amp\Process\Internal\Windows\SocketConnector->onProcessConnectTimeout('bo',
         Object(_HumbugBox5adef95163555\Amp\Process\Internal\Windows\Handle))
#1
         phar:///bin/box.phar/vendor/amphp/amp/lib/Loop/Driver.php(121):
         _HumbugBox5adef95163555\Amp\Loop\NativeDriver->dispatch(true)
#2
         phar:///bin/box.phar/vendor/amphp/amp/lib/Loop/Driver.php(69):
         _HumbugBox5adef95163555\Amp\Loop\Driver->tick()
#3
         phar:///bin/box.phar/vendor/amphp/amp/lib/Loop.php(76):
         _HumbugBox5adef95163555\Amp\Loop\Driver->run()
#4
         phar:///bin/box.phar/vendor/amphp/amp/lib/functions.php(151):
         _HumbugBox5adef95163555\Amp\Loop::run(Object(Closure))
#5 phar:///bin/box.phar/src/Box.php(264):
         _HumbugBox5adef95163555\Amp\Promise\wait(Object(_HumbugBox5adef95163555\Amp\Coroutine))
#6
         phar:///bin/box.phar/src/Box.php(153): _HumbugBox5adef95163555\KevinGH\Box\Box->processContents(Array,
         '...')
#7 phar:///bin/box.phar/src/Console/Command/Compile.php(222):
         _HumbugBox5adef95163555\KevinGH\Box\Box->addFiles(Array, false, true)
#8
         phar:///bin/box.phar/src/Console/Command/Compile.php(140):
         _HumbugBox5adef95163555\KevinGH\Box\Console\Command\Compile->addFiles(Object(_HumbugBox5adef95163555\KevinGH\Bo
         x\Configuration), Object(_HumbugBox5adef95163555\KevinGH\Box\Box),
         Object(_HumbugBox5adef95163555\KevinGH\Box\Console\Logger\BuildLogger),
         Object(_HumbugBox5adef95163555\Symfony\Component\Console\Style\SymfonyStyle))
#9
         phar:///bin/box.phar/src/Console/Command/Compile.php(124):
         _HumbugBox5adef95163555\KevinGH\Box\Console\Command\Compile->createPhar(Object(_HumbugBox5adef95163555\KevinGH
         Box\Configuration), Object(_HumbugBox5adef95163555\Symfony\Component\Console\Input\ArgvInput),
         Object(_HumbugBox5adef95163555\Symfony\Component\Console\Output\ConsoleOutput),
         Object(_HumbugBox5adef95163555\KevinGH\Box\Console\Logger\BuildLogger),
         Object(_HumbugBox5adef95163555\Symfony\Component\Console\Style\SymfonyStyle), false)
#10
         phar:///bin/box.phar/vendor/symfony/console/Command/Command.php(225):
         _HumbugBox5adef95163555\KevinGH\Box\Console\Command\Compile->execute(Object(_HumbugBox5adef95163555\Symfony\Com
         ponent\Console\Input\ArgvInput),
         Object(_HumbugBox5adef95163555\Symfony\Component\Console\Output\ConsoleOutput))
#11
         phar:///bin/box.phar/src/Console/Command/Build.php(34):
         _HumbugBox5adef95163555\Symfony\Component\Console\Command\Command->run(Object(_HumbugBox5adef95163555\Symfony\C
         omponent\Console\Input\ArgvInput),
         Object(_HumbugBox5adef95163555\Symfony\Component\Console\Output\ConsoleOutput))
#12
         phar:///bin/box.phar/vendor/symfony/console/Application.php(753):
         _HumbugBox5adef95163555\KevinGH\Box\Console\Command\Build->run(Object(_HumbugBox5adef95163555\Symfony\Component
         \Console\Input\ArgvInput), Object(_HumbugBox5adef95163555\Symfony\Component\Console\Output\ConsoleOutput))
 ```
</details>

**Comment by theofidry** [comment]
Thanks for reporting the issue!

Amp _is_ optional, do `box compile --no-parallel`.

I didn't think there could be that kind of issue. Maybe we could either process in non-parallel when that happens but then it should be logged IMO. The files processing with PHP-Scoper is _really_ slow, non parallelising this takes ages (easily x4 compilation time).

Unrelated, but you should be able to simplify your config to:

```json
compactors:
   - Herrera\Box\Compactor\Json
   - Herrera\Box\Compactor\Php
compression: GZ
output: build/puppet-enc.phar
git-commit: git_commit
git-version: git_version
git-tag: git_tag
chmod: '0755'
```

Removing useless files like test files & dev dependencies is sorted out by Box. It avoids any extra faff on our end, same for xdebug which is disabled by Box and the autoloader which is optimised.

**Comment by c33s** [comment]
the easierst way to test it for yourself is to install a strict firewall which disallows to  open a port localy.

shouldn't you only have to catch `Amp\Process\ProcessException`?

> could you open an issue on Amp repo regarding this?

should i? i am not sure what to write because i cannot really say if the problem is in how your app calls amp or if amp simply assume that all ports are open everywhere and everytime.

i think the exception handling is the part of box, because amp works per definition with sockets and ports (i assume). so box should handle `if ($connection == false` of `catch Exception...`

for the other stuff i will have to check the amp doc if it is possible or open a ticket at their repo:
- configure which ports are used
- configure the location where workers are stored
- configure the name of the workers

or maybe @kelunik (hope i picked the right maintainer for amphp) has a little time to read through this issue and shares his opinion with us.

**Comment by theofidry** [comment]
I'm not sure which exception to catch yet because it depends on how the exceptions are intercepted and re-thrown.

>should i? i am not sure what to write because i cannot really say if the problem is in how your app calls amp or if amp simply assume that all ports are open everywhere and everytime.
>i think the exception handling is the part of box, because amp works per definition with sockets and ports (i assume). so box should handle if ($connection == false of catch Exception...

Yes. The way Box consumes Amp can't be any simpler: [`wait(parallelMap($files, $processFile))`](https://github.com/humbug/box/blob/master/src/Box.php#L342). So if it doesn't thrown an exception and instead goes in an infinite loop mode it's definitely a bug there

**Comment by c33s** [comment]
oh forgot about the infinit loop xD

**Comment by theofidry** [comment]
Closing in favour of #193

**Comment by c33s** [comment]
shouldn't this be reopend, is it really a windows only? isn't it also about allowing to enable/disable multi threading?

**Comment by c33s** [comment]
`.\bin\box.bat compile --no-parallel`
